consensus/src/msg_handler.rs

```diff
@@ -66,7 +66,33 @@ pub trait MsgHandler {
                 // it is valid or not.
                 self.verify(msg, iteration, round_committees)
             }
-            Status::Future => Err(ConsensusError::FutureEvent),
+            Status::Future => {
+                // Pre-verify future messages for the current round
+                if msg.header.round == ru.round {
+                    if msg_tip != ru.hash() {
+                        return Err(ConsensusError::InvalidPrevBlockHash(
+                            msg_tip,
+                        ));
+                    }
+
+                    let future_committee = round_committees
+                        .get_committee(msg.get_step())
+                        .expect("committee to be created before run");
+
+                    // Ensure the message originates from a committee member.
+                    if !future_committee.is_member(signer) {
+                        return Err(ConsensusError::NotCommitteeMember);
+                    }
+
+                    // Delegate message final verification to the phase
+                    // instance. It is the phase that knows
+                    // what message type to expect and if it
+                    // is valid or not.
+                    self.verify(msg, iteration, round_committees)?;
```

Comment:
shouldn't this return `OK(())` if `self.verify` succeeds? It seems to me that the `Err(ConsensusError::FutureEvent)` is always returned

Reply:
Ah, I get it. We need to signal to the consensus that nothing else should be done as the message is for a future iteration. In this case this is not a mistake (although I consider it pretty ugly to control the flow of the consensus using errors :/)